### PR TITLE
add updated remp doc

### PR DESCRIPTION
**Pull request by a-zorina**

### CODE REVIEW

**Review by eternalflow — APPROVED** [review]

**Review by elasticLove1 — CHANGES_REQUESTED** [review]

File: src/arch/remp.md [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+REMP is a protocol that adds some additional guarantees/features for external messages processing:
+
+**1. Replay protection**
+
+   If a message is processed and added into an accepted block, then the same message **(that is, the message with the same hash)** will not be collated for some time period. If the message has some expiration time (corresponding to the time period), then this effectively makes efficient replay protection.
+
+**2. A message may be sent only once**
+
+If there will be a possibility to accept it and add it to a block, then it will be done. No messages are lost — except for blockchain overloading reasons.
```
[the remark this answers is not in the document]
Comment (a-zorina) [reply src/arch/remp.md]: rephrased the paragpaph

File: src/arch/remp.md [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+**3. One can trace the message processing.**
+
+There are several checkpoints on the message processing path (when validators received the message, when message was added to a block, when the block was finalized, etc) — and depending on the message importance one may trade efficiency for reliability in its software. The authors believe that most messages could be considered processed when validators acknowledge that messages were received (this happens in 100-200 ms; after that it’s highly unlikely that the message is declined). On the other hand, if your transaction is really important, then you can wait till the block with the transaction result is issued.
```
[the remark this answers is not in the document]
Comment (a-zorina) [reply src/arch/remp.md]: rephrased

File: src/arch/remp.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+REMP is a set of protocols and data structures that are designed to keep trace of incoming external messages. Message flow in REMP can be seen on the following diagram:
+
+![](remp1.png)
+
+A message from the user application is sent to a REMP Client (e.g., it could be some Full Node; also the user application may implement the necessary protocol by itself). The client in turn sends the message to its shard validators. The validators, aside from processing the message, gather info about message validation status and send it back to the REMP Client and then further to the user (such messages are called “receipts”).
+
+### REMP Catchain
+
+The validators exchange info between them using special REMP Catchain protocol. Each validator Catchain session (protocol/data structures used for controlling consensus on block candidates) is accompanied by a REMP Catchain session (Reliable Message Queue, RMQ), which is used to store incoming messages. When validator Catchain sessions are started/stopped/split/merged/etc, the same happens with related REMP Catchain session. When the validator session expires, all messages being processed are transferred to the next REMP Catchain sessions.
```
[the remark this answers is not in the document]
Comment (a-zorina) [reply src/arch/remp.md]: fixed

File: src/arch/remp.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+Each validator tries to collate the message until its collator rejects the attempt. 
+
+So, the message processing may result in three outcomes:
+
+- **accepted by collator** - and then validated - and included into shard block - and the block is included into masterchain (accept, finalized)
+- **rejected by fullnode before processing** - or by all collators (reject)
+- **timeout** - its collation/validation attempts were fruitless, and the message became obsolete.
+
+This gives hard replay protection guarantees (if message has expiration mechanism included — making the message obsolete before two catchains are switched), and soft processing guarantees (message collation/validation attempts may be fruitless if the message is wrong, or if the blockchain is overloaded).
```
[the remark this answers is not in the document]
Comment (a-zorina) [reply src/arch/remp.md:61]: added replay protection section

File: src/arch/remp.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+				// shard block the message was included to
+        "block_id": "1692345dc4bd57762330e454926a81dcb45626b433121c51804075b3cc26984e",
+				"block_file_hash": "00a2345dc4bd57762330e454926a81dcb45626b433121c51804075b3cc26984e",
+        "block_seqno": 123,
+        "shard": "3800000000000000",
+        "wc": 0,
+    }
+}
+```
+
```
[the remark this answers is not in the document]
Comment (a-zorina) [reply src/arch/remp.md:224]: added

File: src/arch/remp.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (elasticLove1) [reply src/arch/remp.md:61]: Very good explanation!

**Review by elasticLove1 — APPROVED** [review]

**Pull request merged**